docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1700 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #85.            min_dist = min(min_dist, np.max(approx_P[include]))
</pre><p>It feels a bit weird to be changing <code>min_dist</code> some many times and then using it inside of a <code>for</code> loop later. It feels like <code>min_dist</code> , which is an input parameter is being abused    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Reply:
I agree. I felt a little confused when I was reading that part of code before. one simple/lazy idea is to use a different name For instance, we can use the name `min_dist_init` for input. 
Then:
```
min_dist_init = min(min_dist_init, np.max(approx_P[include]))

min_distances = min_dist_init * (1 - np.arange(max_attempts[i]) * decay[i])
min_distances = np.clip(min_distances, a_min=0.0, a_max=None)
        
for min_dist in min_distances:
    # some code #
    if np.any(is_cands):
         min_dist_init = min_dist
         break

```

I will try to see if I can come up with a better solution.

Reply:
<p>I would be less focused on the name and more focused on the overall design. Ultimately, the name should be <code>min_dist</code> given that it is the name being used elsewhere </p>

Reply:
How do you feel about this design? (notice lines with comment `# NEW`) 

**We add `approx_P` (approx matrix profile) to `_find_candidates`  function:**
```
def _find_candidates(T, m, M_T, Σ_T, min_dist, appox_P, init_cands = None, right=True, finite=False):
           # some code
           for start, stop in cand_idx_chunks:
                QT = core._sliding_dot_product(T[i:i+m], T[start:stop]) 
                D = core._mass(T[i:i+m], T[start:stop], QT, M_T[i], Σ_T[i], M_T[start:stop-m+1], Σ_T[start:stop-m+1])
                appox_P[start:stop-m+1] = np.minimum(appox_P[start:stop-m+1], D) # NEW
                
                mask = np.flatnonzero(D < min_dist)   
                is_cands[start:stop-m+1][mask] = False

                if len(mask):
                    appox_P[i] = min(appox_P[i], D.min()) # NEW
                    is_cands[i] = False
        
    return is_cands, appox_P # returning updated approx_P as well
```

**Why? there are two advantages here:**
1- The `approx_P[is_cands]` (where `is_cands` is the finalized candidates considering both right and left neighbors) is actually TRUE matrix profile values (!) in those indices! So, we can easily bypass the `for-loop` in `_find_discords` that tries to find NN of each of those finalized candidates.

2- We may remove `min_dist` input from `_discords`. 

```
min_dist = np.nanmax(approx_P[include & np.ifinite(approx_P)]) # NEW 
min_distances = min_dist * (1 - np.arange(max_attempts[i]) * decay[i])
min_distances = np.clip(min_distances, a_min=0.0, a_max=None)
```
and,

Reply:
@seanlaw 
Note that the idea of MERLIN was to use top-1 `discord_dist` of subsequence length `m` as `min_dist` of `m+1`. So, we might need to use `discord_dist` of length `m` once, and then we just need to use `approx_P` as I showed above to find top-k discords of a specified length. I will think about the top-level function `discord` and see if I should revise my proposed design.

Reply:
<p>It sounds reasonable. I don't like the idea of mutating an input parameter though as it causes unintended consequences for the user. We should always leave the user's inputs unchanged. We could simply make a copy in those cases if it is only happening once like:</p><p> </p><pre class="ql-syntax" spellcheck="false">approxy_P = approx_P.copy() 
</pre>

Reply:
<p>I think that before we change it, we should move the current code to <code>discords.py</code> and write our unit tests. This way, we can test our changes along the way and confirm that it produces the same results. Otherwise, our changes are significant enough that it may be dangerous.</p>

Reply:
> think that before we change it, we should move the current code to `discords.py` and write our unit tests.

Yes. because the proposed changes affect all private functions... we should  better do unit test. (I will continue thinking to see if there is another way to handle `min_dist` issue) 

Now, I will go through top-level function `discords`.